File: nodejs-soa/inventory-service/src/infrastructure/web/InventoryController.ts
import { Router, Request, Response } from 'express';
import { IInventoryService } from '../../domain/service/IInventoryService';
import { InventoryResponse } from './dto';
import { InventoryItem } from '../../domain/model/InventoryItem';

/**
 * Inventory REST Controller - SOA Service Endpoint.
 * DIP: Depends on IInventoryService interface, not implementation.
 */
export class InventoryController {
  public readonly router: Router;

  constructor(private readonly service: IInventoryService) {
    this.router = Router();
    this.setupRoutes();
  }

  private setupRoutes(): void {
    this.router.post('/', this.create.bind(this));
    this.router.get('/', this.getAll.bind(this));
    this.router.get('/:productId', this.getByProductId.bind(this));
    this.router.patch('/:productId/increase', this.addStock.bind(this));
  }

  private async create(req: Request, res: Response): Promise<void> {
    try {
      const { productId, productName, quantityAvailable } = req.body;
      if (!productId || !productName || quantityAvailable === undefined) {
        res.status(400).json({ error: 'Missing required fields' });
        return;
      }
      const item = await this.service.createItem(productId, productName, quantityAvailable);
      res.status(201).json(this.toResponse(item));
    } catch (error: any) {
      res.status(500).json({ error: error.message });
    }
  }

  private async getAll(req: Request, res: Response): Promise<void> {
    const items = await this.service.getAll();
    res.json(items.map(this.toResponse));
  }

  private async getByProductId(req: Request, res: Response): Promise<void> {
    const item = await this.service.getByProductId(req.params.productId);
    if (!item) { res.status(404).json({ error: 'Not found' }); return; }
    res.json(this.toResponse(item));
  }

  private async addStock(req: Request, res: Response): Promise<void> {
    try {
      const quantity = parseInt(req.query.quantity as string);
      if (!quantity || quantity <= 0) {
        res.status(400).json({ error: 'Quantity must be positive' });
        return;
      }
      const item = await this.service.addStock(req.params.productId, quantity);
      res.json(this.toResponse(item));
    } catch (error: any) {
      res.status(400).json({ error: error.message });
    }
  }

  private toResponse(item: InventoryItem): InventoryResponse {
    return {
      id: item.id,
      productId: item.productId,
      productName: item.productName,
      quantityAvailable: item.quantityAvailable,
      quantityReserved: item.quantityReserved,
      updatedAt: item.updatedAt.toISOString(),
    };
  }
}

File: java-soa/inventory-service/src/main/java/com/ecommerce/inventory/infrastructure/web/InventoryController.java
package com.ecommerce.inventory.infrastructure.web;

import com.ecommerce.inventory.domain.model.InventoryItem;
import com.ecommerce.inventory.domain.service.InventoryService;
import com.ecommerce.inventory.infrastructure.web.dto.CreateInventoryRequest;
import com.ecommerce.inventory.infrastructure.web.dto.InventoryResponse;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.UUID;

/**
 * Inventory REST Controller - SOA Service Endpoint.
 *
 * <p>In SOA, each service exposes a well-defined interface (contract).
 * This controller defines the HTTP contract for the Inventory Service.</p>
 */
@RestController
@RequestMapping("/api/v1/inventory")
public class InventoryController {

    private final InventoryService inventoryService;

    public InventoryController(InventoryService inventoryService) {
        this.inventoryService = inventoryService;
    }

    @PostMapping
    public ResponseEntity<InventoryResponse> addInventory(@Valid @RequestBody CreateInventoryRequest request) {
        InventoryItem item = inventoryService.createInventoryItem(
                request.productId(), request.productName(), request.quantityAvailable());
        return ResponseEntity.status(HttpStatus.CREATED).body(toResponse(item));
    }

    @GetMapping
    public ResponseEntity<List<InventoryResponse>> listAll() {
        List<InventoryResponse> items = inventoryService.getAllInventory()
                .stream().map(this::toResponse).toList();
        return ResponseEntity.ok(items);
    }

    @GetMapping("/{productId}")
    public ResponseEntity<InventoryResponse> getByProductId(@PathVariable UUID productId) {
        return inventoryService.getInventoryByProductId(productId)
                .map(item -> ResponseEntity.ok(toResponse(item)))
                .orElse(ResponseEntity.notFound().build());
    }

    @PatchMapping("/{productId}/increase")
    public ResponseEntity<InventoryResponse> increaseStock(
            @PathVariable UUID productId, @RequestParam int quantity) {
        InventoryItem item = inventoryService.addStock(productId, quantity);
        return ResponseEntity.ok(toResponse(item));
    }

    private InventoryResponse toResponse(InventoryItem item) {
        return new InventoryResponse(
                item.getProductId(), item.getProductName(),
                item.getQuantityAvailable(), item.getQuantityReserved(),
                item.getUpdatedAt());
    }
}
